File: totaleconomy-core/src/main/java/com/ericgrandt/totaleconomy/model/TECurrency.java
package com.ericgrandt.totaleconomy.model;

import net.kyori.adventure.text.Component;

import java.math.BigDecimal;
import java.math.RoundingMode;

public record TECurrency(
    String code, String name, String pluralName, String symbol, int fractionalDigits, boolean isDefault
) implements Currency {
    @Override
    public Component format(BigDecimal amount) {
        var balance = amount.setScale(fractionalDigits, RoundingMode.DOWN);

        if (symbol == null) {
            var suffix = balance.compareTo(BigDecimal.ONE) == 0 ? name : pluralName;
            return Component.text("Balance: %s %s".formatted(balance.toPlainString(), suffix));
        }

        return Component.text("Balance: %s%s".formatted(symbol, balance.toPlainString()));
    }
}

File: totaleconomy-core/src/main/kotlin/com/ericgrandt/totaleconomy/model/TECurrency.kt
package com.ericgrandt.totaleconomy.model

import net.kyori.adventure.text.Component
import java.math.BigDecimal
import java.math.RoundingMode

data class TECurrency(
    override val code: String,
    override val name: String,
    override val pluralName: String,
    override val symbol: String?,
    override val fractionalDigits: Int,
    override val isDefault: Boolean,
) : Currency {
    // TODO: Test
    override fun format(amount: BigDecimal): Component {
        val balance = amount.setScale(fractionalDigits, RoundingMode.DOWN)

        if (symbol == null) {
            val suffix = if (balance.compareTo(BigDecimal.ONE) == 0) name else pluralName
            return Component.text("${balance.toPlainString()} $suffix")
        }

        return Component.text("$symbol${balance.toPlainString()}")
    }
}
